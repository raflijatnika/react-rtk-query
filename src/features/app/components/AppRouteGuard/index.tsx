// React
import { memo, useMemo } from 'react'

// React Router DOM
import { useLocation, Navigate } from 'react-router-dom'

const AppRouteGuard = memo(({ children }: { children: JSX.Element }) => {
  // Hook
  const auth_token = false
  const location = useLocation()
  const isAuthRoute = useMemo(() => {
    return location.pathname.includes('auth')
  }, [location.pathname])

  // Check if user inside auth route and not authenticated
  if (isAuthRoute && !auth_token) {
    return children
  }

  // Check if user inside auth route and already authenticated
  if (isAuthRoute && auth_token) {
    return <Navigate to='/' state={{ from: location }} replace />
  }

  if (!auth_token) {
    // Redirect them to the /auth/login page, but save the current location they were
    // trying to go to when they were redirected. This allows us to send them
    // along to that page after they login, which is a nicer user experience
    // than dropping them off on the home page.
    return <Navigate to='/auth/login' state={{ from: location }} replace />
  }

  return children
})

AppRouteGuard.displayName = 'AppRouteGuard'

export { AppRouteGuard }
