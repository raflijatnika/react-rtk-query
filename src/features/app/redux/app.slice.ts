// Redux Toolkit
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

// Constant
import { APP_SLICE_INITIAL_STATE } from '@/features/app/constant/app-slice.constant'

const app = createSlice({
  name: 'app',
  initialState: APP_SLICE_INITIAL_STATE,
  reducers: {
    app_SET_INITIALIZED: (state, { payload }: PayloadAction<boolean>): void => {
      state.app_isInitialized = payload
    },
    app_SET_CONNECTED: (state, { payload }: PayloadAction<boolean>): void => {
      state.app_isConnected = payload
    }
  }
})

// Mutations
export const { app_SET_INITIALIZED, app_SET_CONNECTED } = app.actions

export default app.reducer
