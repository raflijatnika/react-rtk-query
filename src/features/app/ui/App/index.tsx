const App = () => {
  return <h1>Entry Point</h1>
}

export { App }
