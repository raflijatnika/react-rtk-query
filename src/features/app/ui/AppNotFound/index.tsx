const AppNotFound = () => {
  return <h1>Not Found</h1>
}

export { AppNotFound }
