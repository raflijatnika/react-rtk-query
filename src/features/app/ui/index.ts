export * from './App'
export * from './AppEntryPoint'
export * from './AppNotFound'
