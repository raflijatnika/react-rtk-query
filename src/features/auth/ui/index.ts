export * from './AuthLogin'
export * from './AuthRegister'
export * from './AuthResetPassword'
