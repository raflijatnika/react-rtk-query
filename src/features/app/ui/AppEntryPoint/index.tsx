// React
import { Suspense } from 'react'

// Router
import { useRouter } from '@/plugins'

// Redux
import { Provider } from 'react-redux'
import { PersistGate } from 'redux-persist/integration/react'
import { persistor, store } from '@/plugins/redux'

// Styled Components (Global Styles)
import { AppBaseGlobalStyle } from '@/features/app/components'

// i18n
import '@/plugins/i18n'

// Tailwind
import '@/assets/styles/tailwind.css'

// Antd
import '@/assets/styles/antd.less'

const AppEntryPoint = () => {
  // Hook
  const routes = useRouter()

  return (
    <>
      {/* Global Styles */}
      <AppBaseGlobalStyle />

      {/* Redux, Persist, and Router */}
      <Provider store={store}>
        <PersistGate loading={'Redux Loading...'} persistor={persistor}>
          <Suspense fallback={'Loading...'}>{routes}</Suspense>
        </PersistGate>
      </Provider>
    </>
  )
}

export { AppEntryPoint }
