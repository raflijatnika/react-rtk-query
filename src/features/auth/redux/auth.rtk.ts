// Interfaces
import { IAuthAttrsLogin } from '@/features/auth/interfaces/auth-attrs.interface'
import {
  IAuthResponseToken,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

// Rtk
import { emptySplitApi } from '@/features/app/redux/app.rtk'

// Mutations
import { auth_SET_TOKEN } from './auth.slice'

export const authApi = emptySplitApi.injectEndpoints({
  endpoints: builder => ({
    auth_register: builder.mutation<
      IAuthResponseAuthenticatedUser,
      IAuthAttrsLogin
    >({
      query: ({ body }) => ({
        url: '/auth/register',
        method: 'POST',
        body
      })
    }),
    auth_login: builder.mutation<IAuthResponseToken['result'], IAuthAttrsLogin>(
      {
        query: ({ body }) => ({
          url: '/auth/login',
          method: 'POST',
          body
        }),
        transformResponse: (response: IAuthResponseToken) => response.result,
        onQueryStarted: async (payload, { dispatch, queryFulfilled }) => {
          try {
            const response = await queryFulfilled

            dispatch(auth_SET_TOKEN(response.data))
          } catch {
            //
          }
        }
      }
    ),
    auth_me: builder.query<IAuthResponseAuthenticatedUser['result'], void>({
      query: () => ({
        url: `/auth/me`
      })
    })
  }),
  overrideExisting: false
})

export const {
  useAuth_registerMutation,
  useAuth_loginMutation,
  useAuth_meQuery
} = authApi
