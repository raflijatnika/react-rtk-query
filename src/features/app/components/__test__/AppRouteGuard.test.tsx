// Test
import { render, act } from '@testing-library/react'

// Redux
import { testRedux_renderWithProviders } from '@/plugins/redux/test.redux'
import { setupStore } from '@/plugins'
import { Provider } from 'react-redux'

// Components
import { AppRouteGuard } from '../AppRouteGuard'
import { AppEntryPoint } from '@/features/app/ui/AppEntryPoint'

// Slice
import { auth_login } from '@/features/auth/redux/auth.api'

// React Router DOM
import { MemoryRouter } from 'react-router-dom'

it('AppRouteGuard should render', async () => {
  await act(() => {
    testRedux_renderWithProviders(
      <AppRouteGuard>
        <AppEntryPoint />
      </AppRouteGuard>,
      { useRouting: true }
    )
  })
})

it('AppRouteGuard should have snapshot', async () => {
  await act(() => {
    const { container } = testRedux_renderWithProviders(
      <AppRouteGuard>
        <AppEntryPoint />
      </AppRouteGuard>,
      { useRouting: true }
    )

    expect(container.innerHTML).toMatchSnapshot()
  })
})

it('AppRouteGuard should render content if not authorized', async () => {
  const state = setupStore().getState()

  const route = '/auth/login'
  await act(() => {
    render(
      <MemoryRouter initialEntries={[route]}>
        <Provider store={setupStore()}>
          <AppRouteGuard>
            <AppEntryPoint />
          </AppRouteGuard>
        </Provider>
      </MemoryRouter>
    )
  })

  expect(state.auth.auth_token.token).toEqual(null)
})

it('AppRouteGuard should go to "/" if user already authenticated', async () => {
  const store = setupStore()

  await store
    .dispatch(auth_login({ email: 'john@gmail.com', password: 'password' }))
    .unwrap()

  const route = '/'
  await act(() => {
    render(
      <MemoryRouter initialEntries={[route]}>
        <Provider store={setupStore()}>
          <AppEntryPoint />
        </Provider>
      </MemoryRouter>
    )
  })

  expect(store.getState().auth.auth_token.token).not.toBeNull()
})
