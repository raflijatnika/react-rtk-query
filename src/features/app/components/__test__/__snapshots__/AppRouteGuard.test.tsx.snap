// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRouteGuard should have snapshot 1`] = `""`;
